src/callgraphwidget.cpp

```diff
@@ -0,0 +1,208 @@
+/*
+  callgraphwidget.cpp
+
+  This file is part of Hotspot, the Qt GUI for performance analysis.
+
+  Copyright (C) 2021 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+  Author: Lieven Hey <lieven.hey@kdab.com>
+
+  Licensees holding valid commercial KDAB Hotspot licenses may use this file in
+  accordance with Hotspot Commercial License Agreement provided with the Software.
+
+  Contact info@kdab.com if any conditions of this licensing are not clear to you.
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 2 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "callgraphwidget.h"
+
+#include <QLabel>
+#include <QMouseEvent>
+#include <QSpinBox>
+#include <QTemporaryFile>
+#include <QVBoxLayout>
+
+#include <KColorScheme>
+#include <KParts/ReadOnlyPart>
+#include <KService>
+
+#include "callgraphgenerator.h"
+#include "hotspot-config.h"
+#include "models/callercalleemodel.h"
+#include "settings.h"
+
+#include <kgraphviewer/kgraphviewer_interface.h>
+
+CallgraphWidget::CallgraphWidget(CallerCalleeModel* model, QWidget* parent)
+    : QWidget(parent)
+    , m_graphFile(new QTemporaryFile(this))
+    , m_callerCalleeModel(model)
+{
+    auto layout = new QVBoxLayout(this);
+    layout->setContentsMargins(0, 0, 0, 0);
+    setLayout(layout);
+
+    KService::Ptr service = KService::serviceByDesktopName(QStringLiteral("kgraphviewer_part"));
```

Comment:
the widget is unusable without the service, as such it should take that as an argument. you may need to add a proper factory then to ensure we don't even try to create the widget when the service isn't available